src/initialize/init.cpp

```diff
@@ -27,6 +28,7 @@ extern "C" void lean_initialize() {
     save_stack_info();
     initialize_util_module();
     consume_io_result(initialize_Init(io_mk_world()));
+    consume_io_result(initialize_Std(io_mk_world()));
```

Comment:
@leodemoura It turns out that before this PR, we did not compile all of the stdlib into `lean` after all. While `-rdynamic` must be used to export all symbols from an executable, one still needs `-Wl,--whole-archive` to make sure that even completely unused symbols, such as those in `Std.Data.BinomialHeap`, are included, which this PR does for `libleanshared.so`. That in turn exposed the bug above: since Lean itself never imports that module, its `initializer` function was never called. When running the new test case, the interpreter would use the native code in leanshared, which would crash because of that.

Reply:
@Kha Is this also true for packages which depend on Lean (i.e.,, should Lake also be built with more than just `-rdynamic` as an extra linker flag on Unix-like systems?)

Reply:
@tydeu I don't think so, we do not use static archives when building user executables: https://github.com/leanprover/lean4/blob/8c5b7d8b51159afecf1bf1dd81574d6cd7f70502/src/lean.mk.in#L8
However, this line completely ignores dependencies, so Lake should do something more clever here anyway. The Nix setup gathers static libraries of all transitive dependencies, but that could lead to some symbols not being available to the interpreter as described above: https://github.com/leanprover/lean4/blob/8c5b7d8b51159afecf1bf1dd81574d6cd7f70502/nix/buildLeanPackage.nix#L160

Reply:
@khas Currently Lake does, in fact, gather all static libraries of dependencies. However, most normal packages do not need to link all symbols into their executable, right? And since Lake itself has no dependencies, that should not be a problem their either. So the only place where `-Wl,--whole-archive` might be needed is in building a package that both has static lib dependencies and is interpreting code, right?

Reply:
Correct.